### ISSUE TITLE
Unconnectable Headless Sessions

### ISSUE THREAD

**Issue by Enverex** [Frooxius/NeosPublic#1417]
I have world that runs fine if you host it normally, but for some reason it's simply refusing to be usable on a headless. It exists, it shows up in the headless user's current session list, it shows up on the headless server itself as live and running, but no-one can join it.

Trying to join as a client will result in a permanent loading message stating that it's doing the initial sync, but it will never progress. Client side shows:
```
11:31:41.487 ( 60 FPS)    Resolving SessionID: 
11:31:41.494 ( 60 FPS)    Joining session: lnl-nat:///S-U-EnverexBot:ClubFAD
11:31:41.495 ( 60 FPS)    NetworkInitStart
11:31:41.496 ( 60 FPS)    Network manager: NetX.LNL_Manager - priority: 0
11:31:41.496 ( 60 FPS)    Network manager: FrooxEngine.SteamNetworkManager - priority: -100
11:31:41.497 ( 60 FPS)    Connecting to: lnl-nat:///S-U-EnverexBot:ClubFAD
11:31:41.498 ( 60 FPS)    Punchthrough attempt: 0
11:31:42.547 ( 60 FPS)    Punchthrough attempt: 1
11:31:43.547 ( 60 FPS)    Punchthrough attempt: 2
11:31:44.547 ( 60 FPS)    Punchthrough attempt: 3
11:31:45.547 ( 60 FPS)    Punchthrough attempt: 4
11:31:46.548 ( 60 FPS)    NAT Punchthrough failed, Connecting to Relay
11:31:46.670 ( 60 FPS)    LNL_Connection OnPeerConnected: 52.165.32.220:12600, IsOpen: False
```

But the server never sees this connection, there is absolutely nothing in the headless logs that this connection was ever even attempted. Trying to join other sessions simultaneously hosted on the same headless works fine (and they don't require a relay either, so the relay detection above is also wrongly kicking in).

Other users (e.g. Readun) have also reported the same issue with worlds hosted on headless.

Here's an example of connecting to another session on the same headless which was running at the same time, as you can see it goes through fine and doesn't require a relay either.

```
11:36:51.674 ( 60 FPS)    Resolving SessionID: 
11:36:51.674 ( 60 FPS)    Joining session: lnl-nat:///S-U-EnverexBot:DarkCity
11:36:51.675 ( 60 FPS)    NetworkInitStart
11:36:51.675 ( 60 FPS)    Network manager: NetX.LNL_Manager - priority: 0
11:36:51.675 ( 60 FPS)    Network manager: FrooxEngine.SteamNetworkManager - priority: -100
11:36:51.676 ( 60 FPS)    Connecting to: lnl-nat:///S-U-EnverexBot:DarkCity
11:36:51.676 ( 60 FPS)    Punchthrough attempt: 0
11:36:51.859 ( 60 FPS)    NAT INTRODUCTION SUCCESS: 116.202.173.158:60645, token: s-u-enverexbot:darkcity
11:36:51.871 ( 60 FPS)    NETWORK RECEIVE UNCONNECTED: 116.202.173.158:60645, token: s-u-enverexbot:darkcity, connection token: S-U-EnverexBot:DarkCity
11:36:51.871 ( 60 FPS)    NETWORK RECEIVE UNCONNECTED: 116.202.173.158:60645, token: s-u-enverexbot:darkcity, connection token: S-U-EnverexBot:DarkCity
11:36:51.892 ( 60 FPS)    LNL_Connection OnPeerConnected: 116.202.173.158:60645, IsOpen: False
11:36:51.893 ( 60 FPS)    Connected to: lnl-nat:///S-U-EnverexBot:DarkCity
11:36:51.898 ( 60 FPS)    Waiting for join grant
11:36:52.241 ( 60 FPS)    Join Granted, Begin Data Model Init
11:36:56.701 ( 60 FPS)    PreserveWithAssets: FrooxEngine.AtlasInfo
11:36:57.691 ( 60 FPS)    Starting running world: Dark City```

Not sure what is special about the problem world(s) here.

**Comment by Frooxius** [comment]
Where is the headless server on the network relative to the client you're connecting to it with?

**Comment by Enverex** [comment]
Multiple clients tried. Myself on a residential VDSL connection in the UK, a friend in Finland, someone else in the US, etc. Everyone on standard connections. The headless is a dedicated server in a datacentre in Germany, no special routing should ever be needed for it.

**Comment by Frooxius** [comment]
Does it have a public IP or is it behind NAT? If it's dedicated server, it's odd that it's trying to go through the relay in the first place. 

Is this server typically running at all time so I can check its status and what it reports to the system?

**Comment by Enverex** [comment]
The dedicated server has no natting or IP/port translation of any form, it's what you'd expect on bare-metal dedicated server hosting. The server is up 24/7 because it's what I use for all my websites and email as well as hosting Neos sessions. The IP is 116.202.173.158 (xnode.org) as you can see in the successful session connection in the second block of log text.

**Comment by Frooxius** [comment]
I've just tried connecting to it and it worked fine on my end, a direct connection. Is this problem still persisting for the world? Or does it only occur sometimes?

**Comment by Enverex** [comment]
It's intermittent based on the session itself. It's almost like the session dies and is no-longer joinable. I think in your case I'd restarted the server to update it after the new update was pushed, so the session was working in that instance. But what will happen is that they seem to "die" and become unjoinable at some point until you restart that session on the headless.

**Comment by Frooxius** [comment]
I think I found out what was causing it. When the world restarts, it gets allocated a new port on y our system. Sometimes the world restarts faster than the info of the old session expires and starts announcing itself. The NAT punchthrough server doesn't update the port, but it keeps the session info alive, but any connection attempts will try to use the old port.

I've pushed a fix to the cloud already, see if this issue gets resolved with this.
